File: day5/cpp/fertilizing.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>
#include <sstream>
#include <vector>

using namespace std;

vector<vector<long> > mapVals(fstream& f) {
    string line;
    long src, dest, rlen;
    vector<vector<long> > v;
    int i=0;
    while(f.is_open() && getline(f, line) && !line.empty()) {
        stringstream ss (line);
        ss >> dest;
        ss >> src;
        ss >> rlen;

        v.push_back({src, src+rlen, dest});
        // cout << v[i][0] << endl;
        // i++;
    }

    return v;
}

long checkSeed(long seed, vector<vector<vector<long> > >& m) {
    int i, j;

    i = 0;
    while(i < m.size()) {
        j = 0;
        while (j < m[i].size()) {
            // cout << m[i][j][0] << " " << m[i][j][1] << " " << m[i][j][2] << endl;
            if (seed >= m[i][j][0] && seed < m[i][j][1])
            {
                seed = m[i][j][2] + (seed - m[i][j][0]);
                cout << "START: " << m[i][j][0] << " END: " << m[i][j][1] <<  " Match: " << seed << endl;
                break;
            }
            j++;
        }
        i++;
    }

    cout << endl;
    return seed;
}

int main() {
    fstream f("../input.txt");
    string line, word;
    long tmp, min = LONG_MAX;
    vector<long> seeds;
    vector<vector<vector<long> > > m;
    
    while(f.is_open() && getline(f, line)) {

        stringstream ss (line);
        ss >> word;

        if(word == "seeds:") {
            while (ss >> word) {
                seeds.push_back(stol(word));
            }
        } else if (!line.empty()) {
            // cout << line << endl;
            m.push_back(mapVals(f));
        }
    }

    f.close();

    for (long seed : seeds) {
        cout << "Start: " << seed << endl;
        tmp = checkSeed(seed, m);
        min = tmp < min ? tmp : min;
    }

    cout << "Min value of locations is: " << min << endl;

    return 0;
}

File: week1/day5/fertilizing.py
def mapVals(f):
    line = f.readline()
    v = []
    while True:
      line = f.readline()
      if len(line) < 2: break
      v.append([int(n) for n in line.split()])
      
    return v

def checkSeed(seed, rng, m, i):
  i = 0
  nmin = 0

  if i == len(m): return seed
  
  min = -1
  
  for r in m[i]:
    print(r)
    if seed >= r[1] and seed < (r[1] + r[2]) and (seed + rng) >= (r[1] + r[2]):
        nmin = checkSeed(r[0] + r[2], rng, m, i+1)
        print("1:", nmin)
        if (min < 0 or nmin < min):
            min = nmin
    elif seed >= r[1] and (seed + rng): 
        nmin = checkSeed(r[0] + (seed - r[1]), rng, m, i+1)
        print("2:", nmin)
        if (min < 0 or nmin < min):
            min = nmin
        
  print(min)
  return min


def main():
    m = []
    line = "blah"
    with open('input.txt', 'r') as f:
        while line:
            line = f.readline()
            if (len(line) < 2): 
              continue
            arr = line.split()
            if arr[0] == "seeds:":
                seeds = [int(n) for n in arr[1:len(arr)]]
            else:
                m.append(mapVals(f))
    
    f.close()
    
    min = -1
    
    # print(m)
    
    for i in range(0, len(seeds)-1, 2):
        nmin = checkSeed(seeds[i], seeds[i+1], m, 0)
        if min < 0 or nmin < min: 
            min = nmin
            
    print ("Min found: ", min) 

    
if __name__ == "__main__":
    main()